### PR TITLE
Custom Strategy

### PR DESCRIPTION
**Pull request by zaccharieramzi**
This should allow the evaluation without the unwanted effects of the teardown that removes the model from gpu.

@tomMoral : I don't have a GPU at hand, so can't test this, but it at least runs correctly on CPU.

### CODE REVIEW

File: utils/torch_solver.py [hunk 1]
```diff
@@ -34,7 +34,7 @@ def run(self, callback):
         # Setup the trainer
         trainer = Trainer(
             max_epochs=-1, callbacks=[BenchoptCallback(callback)],
-            accelerator="gpu" if torch.cuda.is_available() else None
+            accelerator="auto",
```
Comment (zaccharieramzi) [inline utils/torch_solver.py:37]: this looks slicker!

File: utils/torch_helper.py [hunk 2]
```diff
@@ -49,3 +58,44 @@ def test_step(self, batch, batch_idx):
...
     def training_step(self, batch, batch_idx):
         return self.loss_logits_y(batch)[0]
+
+
+class DDPStrategyNoTeardown(DDPStrategy):
+    # from
+    # https://pytorch-lightning.readthedocs.io/en/1.6.0/_modules/pytorch_lightning/strategies/ddp.html#DDPStrategy.teardown
+    def teardown(self):
+        log.detail(f"{self.__class__.__name__}: tearing down strategy")
+        super().teardown()
```
Comment (tomMoral) [inline utils/torch_helper.py:68]: As you call this `teardown` I think the model i off-loaded no?

Comment (zaccharieramzi) [reply utils/torch_helper.py:68]: Good call. Can we do super of the grand parent class ?

Comment (zaccharieramzi) [reply utils/torch_helper.py]: I tried the solution from here: https://stackoverflow.com/a/48317309/4332585

File: utils/torch_helper.py [hunk 3]
```diff
@@ -49,3 +55,30 @@ def test_step(self, batch, batch_idx):
...
+        super(SingleDeviceStrategy, self).teardown()
+
+        if self.root_device.type == "cuda":
+            # Do not move the model out
+            log.detail(f"{self.__class__.__name__}: !not! moving model to CPU")
+            # clean up memory
+            torch.cuda.empty_cache()
+
+
+# Register the DDP Strategy with your custom CheckpointIO plugin
```
Comment (zaccharieramzi) [inline utils/torch_helper.py:79]:
```suggestion
# Register the SingleDeviceStrategyNoTeardown Strategy
```

File: utils/torch_solver.py [hunk 4]
```diff
@@ -34,7 +39,7 @@ def run(self, callback):
         # Setup the trainer
         trainer = Trainer(
             max_epochs=-1, callbacks=[BenchoptCallback(callback)],
-            accelerator="gpu" if torch.cuda.is_available() else None
+            accelerator="auto", devices=1
```
Comment (zaccharieramzi) [inline utils/torch_solver.py:42]: here we'll probably need to have a nice way to check the number of devices

**Comment by zaccharieramzi** [comment]
@tomMoral nice one!

File: utils/torch_helper.py [hunk 5]
```diff
@@ -49,3 +55,30 @@ def test_step(self, batch, batch_idx):
...
+            device = "cuda:0"
+        super().__init__(device, accelerator, checkpoint_io, precision_plugin)
+
+    def teardown(self):
+        # Call grand parent teardown
+        super(SingleDeviceStrategy, self).teardown()
+
+        if self.root_device.type == "cuda":
+            # Do not move the model out
+            log.detail(f"{self.__class__.__name__}: !not! moving model to CPU")
```
Comment (tomMoral) [inline utils/torch_helper.py:74]: I think this is not necessary.
```suggestion
```

File: utils/torch_helper.py [hunk 6]
```diff
@@ -1,12 +1,18 @@
...
     import torch
     from torch.nn import functional as F
 
     from torchmetrics import Accuracy
     from pytorch_lightning import LightningModule
     from pytorch_lightning.callbacks import Callback
+    from pytorch_lightning.strategies import SingleDeviceStrategy
+    from pytorch_lightning.strategies import StrategyRegistry
+
+log = logging.getLogger(__name__)
```
Comment (tomMoral) [inline utils/torch_helper.py:15]:
```suggestion
from benchopt import safe_import_context

with safe_import_context() as import_ctx:

    import torch
    from torch.nn import functional as F

    from torchmetrics import Accuracy
    from pytorch_lightning import LightningModule
    from pytorch_lightning.callbacks import Callback
    from pytorch_lightning.strategies import SingleDeviceStrategy
    from pytorch_lightning.strategies import StrategyRegistry
```

File: utils/torch_solver.py (hunk 4 above)
Comment (tomMoral) [reply utils/torch_solver.py:42]: We can see this later on.

**Review by tomMoral — APPROVED** [review]
A few more comments bu in favor of merging this one as is to move on.

**Pull request merged**